### ISSUE TITLE
Some shells don't remember history

### ISSUE THREAD

**Issue by Sampath-Lokuge** [microsoft/vscode#151105]
Does this issue occur when all extensions are disabled?: Yes

- VS Code Version:
- OS Version:

Steps to Reproduce:

1. Open VS code
2. Please see the below images.

![vs1](https://user-images.githubusercontent.com/5206526/171648056-18d407a0-4661-4dc3-af9f-2ec3c6c8de61.png)

When I pressed the top arrow key It shows like so. But no idea from where that CLI is coming. I would like to have `npx cap open android` CLI here.

![vs2](https://user-images.githubusercontent.com/5206526/171648628-8715e879-7b5a-4b09-9a04-da9f5bd75720.png)

**Comment by Tyriar** [comment]
This type of issue has been reported a few times and I think we've closed them all out, let's track in this one. This works fine in pwsh, I believe this is because pwsh persists history to disk as you type, whereas git bash does it when the process shuts down. Something about how we're shutting down the terminal processes doesn't let git bash persist to disk.

Maybe the problem code is here where we kill all processes _before_ closing the conpty pty: https://github.com/microsoft/node-pty/blob/1674722e1caf3ff4dd52438b70ed68d46af83a6d/src/windowsPtyAgent.ts#L161-L168

We tweaked what happens here a bunch to try stop processes leaking which is why we go through and kill all processes in `GetConsoleProcessList`. Maybe we could try calling `native.kill()` and then giving it some time to finish before going through and killing the process list?

**Comment by Okeanos** [comment]
As far as [I understand](https://github.com/git-for-windows/git/discussions/4071) the problem at its root for Git bash integration it indeed appears to be related to:

> Something about how we're shutting down the terminal processes doesn't let git bash persist to disk.

However, the VSCode integration of Git Bash has further, compounding issues:

The [Git Bash Windows Terminal Profile Fragment](https://github.com/git-for-windows/build-extra/blob/main/installer/install.iss#L2814-L2824) vs. [VS Code's definition](https://github.com/microsoft/vscode/blob/main/src/vs/platform/terminal/node/terminalProfiles.ts#L199-L211) reveals that VSCode doesn't start the Git Bash as `interactive` shell. Only interactive Bash sessions will even attempt to save the history by default; a login shell is not sufficient for that!

However, as I outlined in the discussion post … that is sadly not enough and the `export PROMPT_COMMAND="history -a"` workaround is the closest thing to a fix for now I know of.

**Comment by Tyriar** [comment]
The situation might be better in newer versions of Windows when https://github.com/microsoft/terminal/pull/14282 ships. We'll probably need to tweak node-pty's process kill logic though.

**Comment by starball5** [comment]
> This type of issue has been reported a few times and I think we've closed them all out

A couple I found by googling "`site:github.com/microsoft/vscode/issues shell closing window history`": #37273, #85831, #114164. All about Git Bash. Also related: [vscode-docs f0876ad](https://github.com/microsoft/vscode-docs/commit/f0876ad16a03635c899d4409e8d278ea8debbeeb), which changed docs to say this is by-design, and to workaround using `export PROMPT_COMMAND='history -a'`.

Today I encountered [a Stack Overflow question about a user experiencing the same issue with a Zsh shell](https://stackoverflow.com/q/75940733/11107541).

I found I that I experience the same issue with a Bash shell on Ubuntu with VS Code 1.77.1 installed as a Snap. My Bash shell in the VS Code interactive terminal is an interactive, non-login shell. Opening a login sub-shell with `bash -l` makes no difference: History is not saved to the histfile when closing the VS Code window. I have _not_ set `export PROMPT_COMMAND='history -a'`.

**Comment by akpircher** [comment]
I have this problem on VSCode in MacOS/Darwin.

**Comment by akpircher** [comment]
This is an older issue, but I'm hoping that this gets fixed reasonably soon. VSCode is practically unusable for me as it is.

**Comment by JasperKelder** [comment]
Same problem here for zsh on MacOS, not always though. Haven't figured out when or why it happens, but sometimes the integrated terminal in VSCode only shows history of the regular terminal.

**Comment by Okeanos** [comment]
Shells generally have different behaviour concerning histories and VSCode simply exacerbates it by not closing them correctly, especially on Windows in particular (afaik).

Bash will generally only save the history for the _last_ interactive (`-i`) login (`--login`) shell that has been properly terminated. Otherwise the `export PROMPT_COMMAND='history -a'` workaround is required, regardless of VSCode or not. See [interactive behavour](https://www.gnu.org/software/bash/manual/bash.html#Interactive-Shell-Behavior-1) in the manual.

In case of ZSH there are [options](https://zsh.sourceforge.io/Doc/Release/Options.html#History) such as setting `SHARE_HISTORY` that may improve the experience beyond what Bash offers natively for example.

To verify what goes wrong on macOS and help debugging efforts you should verify the Terminal setup in VSCode and add that information here.

**Comment by JasperKelder** [comment]
The SHARE_HISTORY option seems to work for me as a fix.

As far as I know I'm using defaults for the terminal. I just checked it with a VSCode Insider download without any plugin or setting changes and I had the same problem there.

Tldr, for now: 
`open ~/.zshrc`
Add:
`setopt SHARE_HISTORY
`

**Comment by sfboss** [comment]
I had this issue and seemed to only impact vscode (terminal in macos commands were saved and subsequently available in vscode later)... the comment mentioned above by Okeanos to add "export PROMPT_COMMAND='history -a'"  (I added this in to ~/.bash_profile) finally seemed to be the thing that made VScode terminal bash sessions remember history regardless of how VSCode was exited and/or which folder I opened, basically making it seem normal again.  Of note, I did have a messy bash_profile and this command was already in there, but cleaning up the garbage and leaving the command revived it.  Mentioning that in case anyone says they have the command already in there,  maybe a troubleshooting step to backup file and leave only that command and work backwards from there. 

I did several other troubleshooting steps like clearing the history in ~/.bash_history so I can't say absolutely that this specifically fixed it but am about 98% sure.  Hope the details help whomever try and resolve.

**Comment by starball5** [comment]
@sfboss `export PROMPT_COMMAND='history -a'` is a documented workaround. see also [my prior comment](https://github.com/microsoft/vscode/issues/151105#issuecomment-1498406268), what states that. As far as I understand, this issue ticket is about fixing the problem instead of using workarounds.

**Comment by sfboss** [comment]
Wouldn’t they appreciate the information to try and resolve the issue?  By
no means am I advocating for people to give up trying to fix it to use a
workaround and I apologize for missing credit on your previous info.

I am honestly asking if I should not add things like this to issues in the
future as this is the first time I ever commented in a long time.  My
thoughts were it would help people needing a workaround while getting it
fixed fully and/or details to help narrow down a full fix.

Let me know if there is some other way to get that info across or what as
I’d like to make meaningful contributions.

**Comment by starball5** [comment]
> Wouldn’t they appreciate the information to try and resolve the issue?

They wrote the docs that state that that is a known workaround. I.e. I'm saying they already know what you're telling them.

**Comment by migalv** [comment]
I'm getting the same issue on Mac M1 Ventura 13.2 with VSCode v1.78.2 and zsh shell. This started happening when I updated VSCode to the latest version.

**Comment by migalv** [comment]
Hello thanks for the update and the PR @meganrogge. Could you help me figure out how I can test it?

**Comment by meganrogge** [comment]
You can install vscode insiders to test it

**Comment by ginolegigot** [comment]
fwiw, in linux, i figured vscode internal terminal created a .zsh_history when the commands from my native terminal were stored in .zhistory.  
Setting an "export HISTFILE=.zhistory" in .zshrc fixed the issue.

**Comment by KaivG** [comment]
What is the workaround for windows with git bash? Typing exit manually seems to work, but it is obviously tedious. Where do I have to set export PROMPT_COMMAND='history -a' on windows for git bash?

**Comment by timmywil** [comment]
@KaivG 

> What is the workaround for windows with git bash? Typing exit manually seems to work, but it is obviously tedious. Where do I have to set export PROMPT_COMMAND='history -a' on windows for git bash?

You should be able to set it in `~/.bash_profile` or `~/.bashrc`. Find your git bash home with `cd ~ && pwd`. If neither exists already, pick one and make it. Then restart the shell.

**Comment by KaivG** [comment]
> @KaivG
> 
> > What is the workaround for windows with git bash? Typing exit manually seems to work, but it is obviously tedious. Where do I have to set export PROMPT_COMMAND='history -a' on windows for git bash?
> 
> You should be able to set it in `~/.bash_profile` or `~/.bashrc`. Find your git bash home with `cd ~ && pwd`. If neither exists already, pick one and make it. Then restart the shell.

Thanks that worked! Neither file existed, so I created `~/.bash_profile` and stored `PROMPT_COMMAND='history -a'` and now the history is kept. Thanks a lot!

**Comment by Amnesiac9** [comment]
Not working for me even after adding that to my .bash_profile file, running commands do not get saved to the history after closing the terminal with VS Code's kill terminal button or closing the window. This is non-intuitive behavior to have within vscode even if it is emulating something on linux.

**Comment by developer55** [comment]
First of all, thanks to @KaivG for guiding me to the following solution.
This works for me, on Windows 10, Git installed and git bash set to VSCODE as default when starting a new terminal. It may not work for you.
SOLUTION:
1) Create a file "~/.bashrc" in C:\Users\myuser\.bashrc or the path where you have the .bash_history file of your Git installation
2) In .bashrc write in the following line:
export PROMPT_COMMAND='history -a'
3) Save the file.
4) Open VSCODE, and inside it open a git bash terminal. A warning message will appear and Git will generate a ~/.bash_profile in C:\Users\myuser with the following content:
#generated by Git for Windows
test -f ~/.profile && . ~/.profile
test -f ~/.bashrc && . ~/.bashrc
  There will also be a .bash_history file that should already be there. It is in that .bash_history where commands written in git bash will be stored when using the GIT bash terminal built into VSCODE.
